### PR TITLE
std: Expand the area of std::fs

### LINKED ISSUE #24796: Tracking issue for std::fs extensions (RFC #1044)
https://github.com/rust-lang/rfcs/pull/1044

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit is an implementation of [RFC 1044](https://github.com/rust-lang/rfcs/pull/1044) which adds additional
surface area to the `std::fs` module. All new APIs are `#[unstable]` behind
assorted feature names for each one.

The new APIs added are:
- `fs::canonicalize` - bindings to `realpath` on unix and
  `GetFinalPathNameByHandle` on windows.
- `fs::symlink_metadata` - similar to `lstat` on unix
- `fs::FileType` and accessor methods as `is_{file,dir,symlink}`
- `fs::Metadata::file_type` - accessor for the raw file type
- `fs::DirEntry::metadata` - acquisition of metadata which is free on Windows
  but requires a syscall on unix.
- `fs::DirEntry::file_type` - access the file type which may not require a
  syscall on most platforms.
- `fs::DirEntry::file_name` - access just the file name without leading
  components.
- `fs::PathExt::symlink_metadata` - convenience method for the top-level
  function.
- `fs::PathExt::canonicalize` - convenience method for the top-level
  function.
- `fs::PathExt::read_link` - convenience method for the top-level
  function.
- `fs::PathExt::read_dir` - convenience method for the top-level
  function.
- `std::os::raw` - type definitions for raw OS/C types available on all
  platforms.
- `std::os::$platform` - new modules have been added for all currently supported
  platforms (e.g. those more specific than just `unix`).
- `std::os::$platform::raw` - platform-specific type definitions. These modules
  are populated with the bare essentials necessary for lowing I/O types into
  their raw representations, and currently largely consist of the `stat`
  definition for unix platforms.

This commit also deprecates `Metadata::{modified, accessed}` in favor of
inspecting the raw representations via the lowering methods of `Metadata`.

Closes https://github.com/rust-lang/rust/issues/24796

### CODE REVIEW

**Comment by alexcrichton** [comment]
r? @aturon 

Note that this should not be merged until the RFC is approved.

File: src/libstd/sys/windows/ext.rs [hunk 1]
```diff
@@ -1,299 +0,0 @@
...
-    /// # fn foo() -> std::io::Result<()> {
-    /// try!(fs::symlink_file("a", "b"));
-    /// # Ok(())
-    /// # }
-    /// ```
-    pub fn symlink_dir<P: AsRef<Path>, Q: AsRef<Path>> (src: P, dst: Q)
-                                                        -> io::Result<()>
-    {
-        sys::fs2::symlink_inner(src.as_ref(), dst.as_ref(), true)
-    }
```
Comment (lambda) [inline src/libstd/sys/windows/ext.rs]: Looks like you missed moving the new symlink APIs when breaking these modules out into separate files.

File: src/libstd/sys/unix/ext.rs [hunk 2]
```diff
@@ -1,336 +0,0 @@
...
-    ///
-    /// # fn foo() -> std::io::Result<()> {
-    /// try!(fs::symlink("a.txt", "b.txt"));
-    /// # Ok(())
-    /// # }
-    /// ```
-    pub fn symlink<P: AsRef<Path>, Q: AsRef<Path>>(src: P, dst: Q) -> io::Result<()>
-    {
-        sys::fs2::symlink(src.as_ref(), dst.as_ref())
-    }
```
Comment (lambda) [inline src/libstd/sys/unix/ext.rs]: Also missed the Unix version of `symlink`

File: src/libstd/sys/windows/fs2.rs [hunk 3]
```diff
@@ -285,27 +309,26 @@ pub fn to_utf16(s: &Path) -> Vec<u16> {
...
+    pub fn accessed(&self) -> u64 { self.to_ns(&self.data.ftLastAccessTime) }
+    pub fn modified(&self) -> u64 { self.to_ns(&self.data.ftLastWriteTime) }
 
-    fn to_ms(&self, ft: &libc::FILETIME) -> u64 {
-        // FILETIME is in 100ns intervals and there are 10000 intervals in a
-        // millisecond.
-        let bits = (ft.dwLowDateTime as u64) | ((ft.dwHighDateTime as u64) << 32);
-        bits / 10000
+    fn to_ns(&self, ft: &libc::FILETIME) -> u64 {
+        (ft.dwLowDateTime as u64) | ((ft.dwHighDateTime as u64) << 32)
```
Comment (retep998) [inline src/libstd/sys/windows/fs2.rs]: Either the result needs to be multiplied by 100 to actually get nanoseconds, or the function needs to be renamed.

File: src/libstd/sys/windows/fs2.rs [hunk 4]
```diff
@@ -323,6 +346,19 @@ impl FilePermissions {
...
+    pub fn is_dir(&self) -> bool {
+        !self.is_symlink() &&
+            self.file_attributes & c::FILE_ATTRIBUTE_DIRECTORY != 0
+    }
+    pub fn is_file(&self) -> bool {
+        !self.is_symlink() && !self.is_dir()
+    }
+    pub fn is_symlink(&self) -> bool {
+        self.file_attributes & libc::FILE_ATTRIBUTE_REPARSE_POINT != 0
+    }
```
Comment (lambda) [inline src/libstd/sys/windows/fs2.rs]: Not all [reparse points](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365503%28v=vs.85%29.aspx) are symlinks on Windows. Reparse points are a mechanism for storing some application defined data, that can be interpreted by a driver to act as a virtual filesystem at that point in the hierarchy, and are used to provide not only symlinks, but also mountpoints, deduplication, etc. See the [list of Microsoft defined reparse point tags](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365511%28v=vs.85%29.aspx) for details. It looks like you need to get the reparse point tag and check whether it is `IO_REPARSE_TAG_SYMLINK`. It looks only the `FindFirstFile`/`FindNextFile` family of functions return the [`WIN32_FIND_DATA`](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365740%28v=vs.85%29.aspx) that contains this information. I think the best option would be to extend the `FileType` struct with something like `reparse_point_tag: Option<libc::DWORD>`, and when constructing it in places where you have that information; that is, in `DirEntry` where we get it from the `WIN32_FIND_DATAW`, and probably also change `lstat` to use `FindFirstFileW` so we can get this information in that call as well.

File: src/libstd/sys/windows/ext.rs (hunk 1 above)
Comment (alexcrichton) [inline src/libstd/sys/windows/ext.rs]: Oops! I had a feeling I shouldn't have ignored that merge conflict.

**Comment by alexcrichton** [comment]
@lambda 

Thanks [for the info](https://github.com/rust-lang/rust/pull/24711#discussion_r28937068)! I'm moving discussion out here as your comment is now collapsed. This actually poses some difficulties and the current solution makes me feel uncomfortable. There are currently two methods of acquiring a `FileType`, one through `DirEntry` and the other through `FileAttr` (speaking in `fs2` terms). The `DirEntry` one knows whether it's a symlink or not (by reading the `dwReserved0` field you mentioned), but the `FileAttr` does not).

There are a few places which we can get a `FileAttr`:

* `DirEntry::metadata` - this is fine as it already knows whether it's a symlink or not
* `File::metadata` - this is problematic as to learn about whether the file is a symlink it requires a syscall
* `lstat` - this is less problematic as we can switch to using `FindFirstFile` as you mentioned, but it is somewhat unnerving that accidental `*` or `?` characters in the path could produce some odd results.
* `stat` - this suffers the same problems as `File::metadata` and `lstat`

Overall this basically means that we need to have an extra syscall or some different semantics (which now makes me worry about `read_dir`...). This is an unfortunate deviation for what may not be that important all the time, so I suppose it suffices to say that I'm glad that these functions are `#[unstable]` for now!

Regardless, though, I think the current implementation is at least "more correct", although I'm not convinced that we actually want it!

**Comment by lambda** [comment]
I think that the latest approach that you've added looks fine, and I think that the abstraction is reasonable; having to spend an extra syscall to find out if a file is a symlink or not doesn't seem unreasonable, especially since you only pay that extra cost for reparse points, which should be fairly rare. And I think your current approach is fine, you're right, there are some concerns using `FindFirstFile` so this way of doing it should be fine.

On the other hand, I'm not a Windows expert; I just noticed the problem since I'd been reading the documentation closely while reviewing rust-lang/rfcs#1044 and writing rust-lang/rfcs#1048. @retep998 do you have any opinion on this, you seem to have the most Windows API experience.

**Comment by retep998** [comment]
As long as we're only doing that extra syscall for reparse points, the cost should be negligible since most files are not reparse points. That said, the implementation of invoking `readlink` just to find out whether a reparse point is a symlink seems a bit too heavy. It would be nice to eventually expose all the possible reparse tag variants, not just symbolic links.

File: src/libstd/fs.rs [hunk 5]
```diff
@@ -674,6 +716,41 @@ impl DirEntry {
...
+    /// symlink.
+    #[unstable(feature = "dir_entry_ext", reason = "recently added API")]
+    pub fn file_type(&self) -> io::Result<FileType> {
+        self.0.file_type().map(FileType)
+    }
+
+    /// Returns the bare file name of this directory entry without any other
+    /// leading path component.
+    #[unstable(feature = "dir_entry_ext", reason = "recently added API")]
+    pub fn file_name(&self) -> OsString {
```
Comment (GuillaumeGomez) [inline src/libstd/fs.rs]: Like I pointed out in the RFC (a little lately, sorry for that). If the entry is a directory or something else, is `file_name` is still a good method name ?

File: src/libstd/sys/windows/fs2.rs [hunk 6]
```diff
@@ -111,8 +120,31 @@ impl DirEntry {
     }
 
     pub fn path(&self) -> PathBuf {
+        self.root.join(&self.file_name())
+    }
+
+    pub fn file_name(&self) -> OsString {
```
Comment (GuillaumeGomez) [inline src/libstd/sys/windows/fs2.rs]: Like I pointed out in the RFC (a little lately, sorry for that). If the entry is a directory or something else, is file_name is still a good method name ?

Comment (alexcrichton) [inline src/libstd/sys/windows/fs2.rs]: This matches the `file_name` methods of `Path`, which was the source for this name. The directory/file distinction here is known to not apply much.

Comment (GuillaumeGomez) [inline src/libstd/sys/windows/fs2.rs]: I'm not really fine with that. I find it not clear. It's really a shame that I was aware of the RFC so lately...

File: src/libstd/fs.rs [hunk 7]
```diff
@@ -674,6 +716,41 @@ impl DirEntry {
     /// The exact text, of course, depends on what files you have in `.`.
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn path(&self) -> PathBuf { self.0.path() }
+
+    /// Return the metadata for the file that this entry points at.
+    ///
+    /// This function may not require an extra system call on some platforms
```
Comment (aturon) [inline src/libstd/fs.rs]: How about: "This function may require an extra system call on some platforms" (or, better yet: we could go ahead and add the `# Platform behavior` section and give a fuller breakdown).

Comment (alexcrichton) [inline src/libstd/fs.rs]: Good idea! I've started out with the platform behavior sections to see where it takes us.

**Comment by alexcrichton** [comment]
@aturon I actually forgot that I didn't implement `DirBuilder`. I've just added a small commit which does so, re-r?

**Comment by aturon** [comment]
@alexcrichton Lovely!

@bors: r+

**Pull request merged**